edb/pgsql/compiler/relctx.py

```diff
@@ -1779,6 +1807,103 @@ def range_for_material_objtype(
     return rvar
 
 
+def _get_typeref_descendants(
+    typeref: irast.TypeRef,
+    *,
+    include_descendants: bool,
+    for_mutation: bool,
+) -> list[irast.TypeRef]:
+    if (
+        include_descendants
+        and not for_mutation
+    ):
+        descendants = [typeref, *irtyputils.get_typeref_descendants(typeref)]
+
+        # Try to only select from actual concrete types.
+        concrete_descendants = [
+            subref for subref in descendants if not subref.is_abstract
+        ]
+        # If there aren't any concrete types, we still need to
+        # generate *something*, so just do the initial one.
+        if concrete_descendants:
+            return concrete_descendants
+        else:
+            return [typeref]
+
+    else:
+        return [typeref]
+
+
+def _selects_for_typeref_descendants(
+    typeref_descendants: Sequence[irast.TypeRef],
+    path_id: irast.PathId,
+    *,
+    for_mutation: bool,
+    include_descendants: bool,
+    ctx: context.CompilerContextLevel,
+) -> list[pgast.SelectStmt]:
+    selects = []
+    for subref in typeref_descendants:
+        rvar = _table_from_typeref(
+            subref,
+            path_id,
+            for_mutation=for_mutation,
+            include_descendants=include_descendants,
+            ctx=ctx,
+        )
+        qry = pgast.SelectStmt(from_clause=[rvar])
+        sub_path_id = path_id
+        pathctx.put_path_value_rvar(qry, sub_path_id, rvar)
+        pathctx.put_path_source_rvar(qry, sub_path_id, rvar)
+
+        selects.append(qry)
+
+    return selects
+
+
+def _table_from_typeref(
+    typeref: irast.TypeRef,
+    path_id: irast.PathId,
+    *,
+    for_mutation: bool,
+    include_descendants: bool,
+    ctx: context.CompilerContextLevel,
+) -> pgast.PathRangeVar:
+    assert isinstance(typeref.name_hint, sn.QualName)
+
+    aspect = 'table'
+
+    if (
+        include_descendants
+        and not for_mutation
+        and typeref.name_hint.module in {'cfg', 'sys', 'schema'}
+    ):
+        aspect = 'inhview'
```

Comment:
@msullivan Thoughts on this? I would like to add a comment here, but I'm not entirely sure how best to explain this branch beyond something vague like, "these modules use views to implement inheritance".